### PR TITLE
security

### PR DESCRIPTION
**Pull request by VasylAzarov**
implemented User model, repo, service, dtos, AuthenticationController, mapper | add user validation | add user registration api

### CODE REVIEW

File: src/main/java/dev/vasyl/proj/controller/AuthenticationController.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        description = "API for managing authentication")
+public class AuthenticationController {
+    private final UserService userService;
+
+    @PostMapping("/registration")
+    public UserResponseDto register(@RequestBody @Valid UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        return userService.register(requestDto);
+    }
+
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/controller/AuthenticationController.java]:
```suggestion
```

File: src/main/java/dev/vasyl/proj/controller/AuthenticationController.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/auth")
+@Tag(name = "Authentication manager",
+        description = "API for managing authentication")
+public class AuthenticationController {
+    private final UserService userService;
+
+    @PostMapping("/registration")
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/controller/AuthenticationController.java:24]: add swagger documentation

File: src/main/java/dev/vasyl/proj/exception/CustomGlobalExceptionHandler.java [hunk 3]
```diff
@@ -44,4 +45,14 @@ private String getErrorMessage(ObjectError e) {
         return e.getDefaultMessage();
     }
 
+    @ExceptionHandler(RegistrationException.class)
+    protected ResponseEntity<Object> handleRegistrationException(
+            RegistrationException ex, WebRequest request) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
+        body.put("status", HttpStatus.CONFLICT);
+        body.put("message", ex.getMessage());
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/exception/CustomGlobalExceptionHandler.java]: You can move it to the private method and use it for other exceptions. For example, EntityNotFoundException should return a 404 response

File: src/main/java/dev/vasyl/proj/service/impl/UserServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/service/impl/UserServiceImpl.java]: better use exists by email ;)

File: src/main/java/dev/vasyl/proj/validation/PasswordValidator.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
+package dev.vasyl.proj.validation;
+
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+
+public class PasswordValidator implements ConstraintValidator<ValidPassword, String> {
+    public static final String PASSWORD_UPPERCASE_REGEX = ".*[A-Z].*";
+    public static final String PASSWORD_DIGIT_REGEX = ".*\\d.*";
+    public static final String PASSWORD_SPECIAL_CHAR_REGEX = ".*[!@#$%^&*(),.?\":{}|<>].*";
+    public static final String PASSWORD_MIN_LENGTH_REGEX = ".{8,}";
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/validation/PasswordValidator.java]: should be private

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+                  constraints:
+                    nullable: false
+              - column:
+                  name: last_name
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: shipping_address
+                  type: varchar(1000)
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/02-create-users-table.yaml]: you should add the same size validation to request dto

File: src/main/java/dev/vasyl/proj/validation/PasswordValidator.java (hunk 5 above)
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/validation/PasswordValidator.java]: do you really need regex for size validation?

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s improve your solution ;)

File: src/main/java/dev/vasyl/proj/dto/user/UserRegistrationRequestDto.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+@ValidPassword(passwordField = "password", confirmPasswordField = "confirmedPassword")
+public class UserRegistrationRequestDto {
+    @Schema(description = "User email",
+            example = "user@email.com")
+    @NotBlank(message = "Email must not be empty")
+    @Email(message = "Invalid email format")
+    private String email;
+    @Schema(description = "User password",
+            example = "Qwerty12345678$")
+    private String password;
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/dto/user/UserRegistrationRequestDto.java:21]:
```suggestion
    @NotBlank
    private String password;
```

File: src/main/java/dev/vasyl/proj/dto/user/UserRegistrationRequestDto.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    @Schema(description = "User confirmed password,"
+            + " mast be similar with password",
+            example = "Qwerty12345678$")
+    private String confirmedPassword;
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/dto/user/UserRegistrationRequestDto.java:26]:
```suggestion
    @NotBlank
    private String confirmedPassword;
```

File: src/main/java/dev/vasyl/proj/validation/PasswordMatchValidator.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+            Field passwordField
+                    = value.getClass().getDeclaredField(this.passwordField);
+            Field confirmPasswordField
+                    = value.getClass().getDeclaredField(this.confirmPasswordField);
+            passwordField.setAccessible(true);
+            confirmPasswordField.setAccessible(true);
+            String password = (String) passwordField.get(value);
+            String confirmPassword = (String) confirmPasswordField.get(value);
+
+            if (password == null || confirmPassword == null) {
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/validation/PasswordMatchValidator.java]: you can use Objects.equals() (minor)

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml (hunk 6 above)
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/02-create-users-table.yaml]: 255 will be enough 

```suggestion
                  type: varchar(255)
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**